### PR TITLE
[DBParameterGroup] Add soft failing for DBParameterGroup

### PR DESCRIPTION
**Pull request by moataz-mhmd**
Introducing access denied soft-fail while updating/reading tags for DBParameterGroup.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: aws-rds-dbparametergroup/src/main/java/software/amazon/rds/dbparametergroup/BaseHandlerStd.java [hunk 1]
```diff
@@ -73,6 +81,62 @@
...
+                    progress,
+                    exception,
+                    SOFT_FAIL_NPROGRESS_TAGGING_ERROR_RULE_SET
+            );
+        }
+
+        return progress;
+    }
+
+    protected CallChain.Completed<DescribeDbParameterGroupsRequest, DescribeDbParameterGroupsResponse, RdsClient, ResourceModel, CallbackContext>
```
Comment (osdrv) [inline aws-rds-dbparametergroup/src/main/java/software/amazon/rds/dbparametergroup/BaseHandlerStd.java:113]: Nit: the return type looked so ugly that I decided to introduce a bit of code duplication.

File: aws-rds-dbparametergroup/src/main/java/software/amazon/rds/dbparametergroup/BaseHandlerStd.java [hunk 2]
```diff
@@ -73,6 +81,62 @@
...
+                        ));
+    }
+
+    protected ProgressEvent<ResourceModel, CallbackContext> fetchDBParameterGroupArn(final AmazonWebServicesClientProxy proxy,
+                                                                                     final ProxyClient<RdsClient> proxyClient,
+                                                                                     final ProgressEvent<ResourceModel, CallbackContext> progress) {
+        return describeDbParameterGroups(proxy, proxyClient, progress.getResourceModel(), progress.getCallbackContext())
+                .done((describeDbParameterGroupsRequest, describeDbParameterGroupsResponse, invocation, resourceModel, context) -> {
+                    final String arn = describeDbParameterGroupsResponse.dbParameterGroups().stream().findFirst().get().dbParameterGroupArn();
+                    context.setDbParameterGroupArn(arn);
```
Comment (osdrv) [inline aws-rds-dbparametergroup/src/main/java/software/amazon/rds/dbparametergroup/BaseHandlerStd.java:135]: I recall your comment about this approach in DBInstance. This is exactly what I pictured it back then. I still haven't decided if I find the context population approach any reliable, let's A/B test it.

File: aws-rds-dbparametergroup/src/main/java/software/amazon/rds/dbparametergroup/CreateHandler.java [hunk 3]
```diff
@@ -16,35 +18,61 @@
...
+        final ResourceModel desiredModel = request.getDesiredResourceState();
+
+        final Tagging.TagSet extraTags = Tagging.TagSet.builder()
+                .stackTags(Tagging.translateTagsToSdk(request.getDesiredResourceTags()))
+                .resourceTags(new HashSet<>(Translator.translateTagsToSdk(request.getDesiredResourceState().getTags())))
+                .build();
+
+        return ProgressEvent.progress(desiredModel, callbackContext)
+                .then(progress -> setDBParameterGroupNameIfEmpty(request, desiredModel, progress))
+                .then(progress -> createDBParameterGroup(proxy, proxyClient, progress, request.getSystemTags(), requestLogger))
```
Comment (osdrv) [inline aws-rds-dbparametergroup/src/main/java/software/amazon/rds/dbparametergroup/CreateHandler.java:39]: I think I would prefer to pack all tags in a single-type container, even though this container will only deliver a set of system tags. This approach will make `createDBParameterGroup` more generic to what exactly came its way.

File: aws-rds-dbparametergroup/src/main/java/software/amazon/rds/dbparametergroup/CreateHandler.java [hunk 4]
```diff
@@ -16,35 +18,61 @@
...
+                .then(progress -> updateTags(proxyClient, progress, Tagging.TagSet.emptySet(), extraTags))
+                .then(progress -> applyParameters(proxy, proxyClient, progress.getResourceModel(), progress.getCallbackContext(), requestLogger))
                 .then(progress -> new ReadHandler().handleRequest(proxy, request, callbackContext, proxyClient, requestLogger));
     }
+
+    private ProgressEvent<ResourceModel, CallbackContext> createDBParameterGroup(final AmazonWebServicesClientProxy proxy,
+                                                                                 final ProxyClient<RdsClient> proxyClient,
+                                                                                 final ProgressEvent<ResourceModel, CallbackContext> progress,
+                                                                                 final Map<String, String> systemTags,
```
Comment (osdrv) [inline aws-rds-dbparametergroup/src/main/java/software/amazon/rds/dbparametergroup/CreateHandler.java:48]: We should abstract it from the specific tag set that we are fancy providing in this case.

File: aws-rds-dbparametergroup/src/main/java/software/amazon/rds/dbparametergroup/CreateHandler.java [hunk 5]
```diff
@@ -16,35 +18,61 @@
...
+                                                                                 final RequestLogger requestLogger) {
+        return proxy.initiate("rds::create-db-parameter-group", proxyClient, progress.getResourceModel(), progress.getCallbackContext())
+                .translateToServiceRequest(resourceModel -> Translator.createDbParameterGroupRequest(resourceModel, systemTags))
+                .backoffDelay(CONSTANT)
```
Comment (osdrv) [inline aws-rds-dbparametergroup/src/main/java/software/amazon/rds/dbparametergroup/CreateHandler.java:52]: I suggest using a config-based delay: `config.getBackoff()`. Here is an example: https://github.com/aws-cloudformation/aws-cloudformation-resource-providers-rds/blob/master/aws-rds-dbinstance/src/main/java/software/amazon/rds/dbinstance/CreateHandler.java#L30

This allows us to override these backoff constants and significantly speed up unit tests probing stabilization.

File: aws-rds-dbparametergroup/src/main/java/software/amazon/rds/dbparametergroup/ReadHandler.java [hunk 6]
```diff
@@ -22,28 +21,40 @@
...
+                .then(progress -> readTags(proxyClient, progress));
     }
+
+    protected ProgressEvent<ResourceModel, CallbackContext> readTags(
+            final ProxyClient<RdsClient> proxyClient,
+            final ProgressEvent<ResourceModel, CallbackContext> progress) {
+        ResourceModel model = progress.getResourceModel();
+        CallbackContext context = progress.getCallbackContext();
+        try {
+            String arn = progress.getCallbackContext().getDbParameterGroupArn();
```
Comment (osdrv) [inline aws-rds-dbparametergroup/src/main/java/software/amazon/rds/dbparametergroup/ReadHandler.java:47]: I think we might need to check its presence here or at least assert it is not null

File: aws-rds-dbparametergroup/src/main/java/software/amazon/rds/dbparametergroup/Translator.java [hunk 7]
```diff
@@ -100,35 +101,38 @@ else if (parameter.applyType().equalsIgnoreCase(ParameterType.Dynamic.toString()
...
+    static ResourceModel translateFromDBParameterGroup(final DBParameterGroup dbParameterGroup) {
         return ResourceModel.builder()
                 .dBParameterGroupName(dbParameterGroup.dbParameterGroupName())
                 .description(dbParameterGroup.description())
                 .family(dbParameterGroup.dbParameterGroupFamily())
-                .tags(tags)
                 .build();
     }
+
+    static List<Tag> translateTags(final Set<software.amazon.awssdk.services.rds.model.Tag> rdsTags) {
```
Comment (osdrv) [inline aws-rds-dbparametergroup/src/main/java/software/amazon/rds/dbparametergroup/Translator.java:129]: nit: I believe I agree with the sentiment of using generic containers. `Collection` could make it more universal.

**Review by osdrv — APPROVED** [review]

File: aws-rds-dbparametergroup/src/main/java/software/amazon/rds/dbparametergroup/ReadHandler.java (hunk 6 above)
Comment (moataz-mhmd) [reply aws-rds-dbparametergroup/src/main/java/software/amazon/rds/dbparametergroup/ReadHandler.java:65]: Done.

File: aws-rds-dbparametergroup/src/main/java/software/amazon/rds/dbparametergroup/Translator.java (hunk 7 above)
Comment (moataz-mhmd) [reply aws-rds-dbparametergroup/src/main/java/software/amazon/rds/dbparametergroup/Translator.java]: Done.

File: aws-rds-dbparametergroup/src/main/java/software/amazon/rds/dbparametergroup/BaseHandlerStd.java (hunk 1 above)
Comment (moataz-mhmd) [reply aws-rds-dbparametergroup/src/main/java/software/amazon/rds/dbparametergroup/BaseHandlerStd.java]: Removed it

**Review by osdrv — APPROVED** [review]

**Review by osdrv — APPROVED** [review]

**Review by dbbh — APPROVED** [review]

**Pull request merged**